source/fuzz/fuzzer_pass_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,164 @@
+          return;
+        }
+
+        // It must be valid to insert an OpCompositeConstruct instruction
+        // before |instruction_iterator|.
+        if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(
+                SpvOpCompositeConstruct, instruction_iterator)) {
+          return;
+        }
+        // Get the scalar type represented by the targeted instruction id.
+        uint32_t scalar_type_id = instruction_iterator->type_id();
+
+        // Get a random vector size from 2 to 4.
+        uint32_t component_count = 2 + std::rand() % 3;
+        // Get the vector type of size range from 2 to 4 of the corresponding
+        // scalar type.
+        uint32_t vec_type_id =
+            FindOrCreateVectorType(scalar_type_id, component_count);
+
+        // Randomly choose a position that target ids should be placed at.
+        // The position is in range [0, n - 1], where n is the size of the
+        // vector.
+        uint32_t position = std::rand() % component_count;
+
+        // target ids are the two scalar ids from the original instruction.
+        uint32_t target_id1 = instruction_iterator->GetSingleWordOperand(1);
+        uint32_t target_id2 = instruction_iterator->GetSingleWordOperand(2);
+
+        // Stores the ids of scalar constants.
+        std::vector<uint32_t> vec1_components;
+        std::vector<uint32_t> vec2_components;
+
+        // Width is specified in the first index for either OpTypeInt or
+        // OpTypeFloat.
+        uint32_t width = GetIRContext()
+                             ->get_def_use_mgr()
+                             ->GetDef(scalar_type_id)
+                             ->GetSingleWordOperand(0);
+        // Get the scalar type.
+        auto type = GetIRContext()->get_type_mgr()->GetType(scalar_type_id);
+
+        // Whether the constant is signed, not used for float type so default
+        // set to true.
+        bool is_signed =
+            type->AsInteger() ? type->AsInteger()->IsSigned() : true;
+
+        // Populate components based on vector type and size.
+        for (uint32_t i = 0; i < component_count; ++i) {
+          if (i == position) {
+            vec1_components.emplace_back(target_id1);
+            vec2_components.emplace_back(target_id2);
+          } else {
+            switch (instruction_iterator->opcode()) {
+              case SpvOpTypeInt: {
+                AddRandomIntConstant(vec1_components, width, is_signed);
+                AddRandomIntConstant(vec2_components, width, is_signed);
+                break;
+              }
+              case SpvOpTypeFloat: {
+                AddRandomFloatConstant(vec1_components, width);
+                AddRandomFloatConstant(vec2_components, width);
+                break;
+              }
+              default:
+                assert(false &&
+                       "Instruction opcode must be a valid numeric constant "
+                       "type.");
+            }
+          }
+        }
```

Comment:
I see what you are doing here. This is a valid approach but it's usually a good idea to reduce the amount of work transformations have to do.

For example, instead of initializing a vector with random components, we can initialize it with 0s and rely on some other fuzzer pass and a transformation to change them in the future. 

This way, if assigning a random value to some particular vector component turns out to be the only transformation triggering the bug, we can remove others, thus providing a smaller and cleaner binary as a bug report to the developers.

Thus, it'd be better to initialize all vector components to some particular value (e.g. 0) using one of `FindOrCreate*` methods. Some things to consider:
- We don't support division instructions ATM but if we decide to do so, we should be careful not to introduce undefined behaviour dividing by 0.
- It might be beneficial to be able to change vector components separately. If we use a single `FindOrCreate` to initialize them, then they will be bound together to a single constant since we don't have the API to create different constants with the same value. Thus, I think you can just increment the value of each component (probably starting from `1` to avoid problems with division by `0` in the future).

TL;DR
Avoid initializing components to random values - better to leave to other transformations. Use one of `FindOrCreate*` methods to create constants for all vector components, initializing all constants to values from the range `[1; n]` sequentially (ignoring target operands or course) where `n` is the size of the vector. See the above explanation for the reasoning behind this.

Reply:
done.